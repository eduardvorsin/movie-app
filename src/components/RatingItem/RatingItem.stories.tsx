import type { Meta, StoryObj } from '@storybook/react';
import RatingItem from './RatingItem';

const meta: Meta<typeof RatingItem> = {
	title: 'UI/RatingItem',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the component'
		},
		children: {
			description: 'The text content of the component',
		},
		isOrdered: {
			description: 'Allows you to add a css counter for a component.'
		},
		testId: {
			description: 'Id for testing the component',
		},
		id: {
			description: 'A unique identifier for the component',
		},
		src: {
			description: 'The path to the image for the component',
		},
		element: {
			description: 'Allows you to determine which html element a component can be, div and li are available.',
		},
		title: {
			description: 'The text for the component header',
		},
		loading: {
			description: 'Allows you to add lazy loading for an image',
		},
	},
	component: RatingItem,
	parameters: {
		docs: {
			description: {
				component: 'A component that allows you to make a list of films for a certain top, for example, the top 10 films with a box office.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof RatingItem>;

export const Default: Story = {
	args: {
		id: 10,
		isOrdered: false,
		element: 'div',
		title: 'Rating Item Title',
		src: '/assets/images/movie-card-placeholder-l-v.svg',
		children: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const Ordered: Story = {
	args: {
		id: 10,
		isOrdered: true,
		element: 'li',
		src: '/assets/images/movie-card-placeholder-l-v.svg',
		children: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.',
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
	render: (args) => (
		<ul>
			<RatingItem
				{...args}
				title='Rating Item Title 1'

			/>
			<RatingItem
				{...args}
				title='Rating Item Title 2'
			/>
		</ul>
	),
};