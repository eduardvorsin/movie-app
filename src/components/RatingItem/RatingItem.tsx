import ThemedImage from "@/components/ThemedImage/ThemedImage";
import Title from "@/components/Title/Title";
import { GeneralProps } from "@/types/shared";
import Link from "next/link";
import { ReactNode } from "react";
import { routes } from "src/constants";

type Props = {
	id: number,
	title: string,
	src: string,
	children: ReactNode,
	element: 'div' | 'li',
	isOrdered?: boolean,
	loading?: 'eager' | 'lazy',
} & GeneralProps;

export default function RatingItem({
	id,
	title,
	src,
	children,
	testId,
	className,
	element,
	isOrdered = true,
	loading,
	...props
}: Props) {
	const classes = [
		'flex items-center gap-4 relative py-2 lg:p-3 [&:nth-child(-n+9)]:border-b-1 [&:nth-child(-n+9)]:border-neutral-300 dark:[&:nth-child(-n+9)]:border-dark-neutral-350 sm:[&:nth-last-child(2)]:border-none transition-colors duration-150 hover:bg-neutral-100 active:bg-neutral-0 dark:hover:bg-dark-neutral-200 dark:active:bg-dark-neutral-250',
		isOrdered ? 'md:[counter-increment:list-item_1] before:hidden md:before:inline-block before:content-[counter(list-item,decimal-leading-zero)] before:text-500 lg:before:text-700 before:font-bold before:text-neutral-1000 dark:before:text-dark-neutral-900 before:leading-none' : '',
		className
	].join(' ');

	const Component = element;

	return (
		<Component
			key={id}
			className={classes}
			data-testid={testId}
			{...props}
		>
			<Link
				className='grow flex items-center gap-3 before:absolute before:w-full before:h-full before:top-0 before:left-0'
				href={`${routes.movies}${id}`}
				title={title}
			>
				<ThemedImage
					className='w-[50px] h-[75px] lg:w-[60px] lg:h-[90px] aspect-[2/3]'
					width={60}
					height={90}
					alt={title}
					src={{
						light: src,
						dark: src,
					}}
					fallback={{
						light: '/assets/images/movie-placeholder-l-v.svg',
						dark: '/assets/images/movie-placeholder-d-v.svg',
					}}
					sizes='60px'
					loading={loading}
				/>

				<Title
					className='text-neutral-1000 dark:text-dark-neutral-900 grow [&]:text-200 md:[&]:text-[1.125rem] line-clamp-3'
					level={5}
					as='h3'
				>
					{title}
				</Title>

				{children}
			</Link>
		</Component>
	);
}