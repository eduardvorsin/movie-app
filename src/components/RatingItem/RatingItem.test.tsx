import { render, screen } from "@testing-library/react";
import RatingItem from "./RatingItem";

describe('RatingItem tests', () => {
	it('is rendered correctly', () => {
		render(
			<RatingItem
				src='/assets/images/movie-card-placeholder-l-v.svg'
				id={0}
				title='test rating item'
				testId='test-rating-item'
				element='div'
			>
				test content
			</RatingItem>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-rating-item')).toBeInTheDocument();
	});

	it('is a basic snapshot', () => {
		render(
			<RatingItem
				src='/assets/images/movie-card-placeholder-l-v.svg'
				id={0}
				title='test rating item'
				testId='test-rating-item'
				element='div'
			>
				test content
			</RatingItem>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-rating-item')).toMatchSnapshot();
	});

	it('a snapshot with isOrdered set to true', () => {
		render(
			<RatingItem
				src='/assets/images/movie-card-placeholder-l-v.svg'
				id={0}
				title='test rating item'
				testId='test-rating-item'
				isOrdered
				element='div'
			>
				test content
			</RatingItem>
		);

		expect(screen.getByTestId<HTMLDivElement>('test-rating-item')).toMatchSnapshot();
	});
});