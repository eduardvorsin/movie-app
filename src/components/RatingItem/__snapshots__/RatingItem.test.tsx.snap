// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingItem tests a snapshot with isOrdered set to true 1`] = `
<div
  class="flex items-center gap-4 relative py-2 lg:p-3 [&:nth-child(-n+9)]:border-b-1 [&:nth-child(-n+9)]:border-neutral-300 dark:[&:nth-child(-n+9)]:border-dark-neutral-350 sm:[&:nth-last-child(2)]:border-none transition-colors duration-150 hover:bg-neutral-100 active:bg-neutral-0 dark:hover:bg-dark-neutral-200 dark:active:bg-dark-neutral-250  md:[counter-increment:list-item_1] before:hidden md:before:inline-block before:content-[counter(list-item,decimal-leading-zero)] before:text-500 lg:before:text-700 before:font-bold before:text-neutral-1000 dark:before:text-dark-neutral-900 before:leading-none before:transition-colors before:duration-150 "
  data-testid="test-rating-item"
>
  <a
    class="grow flex items-center gap-3 before:absolute before:w-full before:h-full before:top-0 before:left-0"
    href="/movies/0"
    title="test rating item"
  >
    <img
      alt="test rating item"
      class="w-[50px] h-[75px] lg:w-[60px] lg:h-[90px] aspect-[2/3]"
      data-nimg="1"
      decoding="async"
      height="90"
      loading="lazy"
      src="http://localhost/assets/images/movie-card-placeholder-l-v.svg"
      style="color: transparent;"
      width="60"
    />
    <h3
      class="leading-[1.25] -tracking-[0.01em] transition-colors duration-150 text-300 font-bold text-neutral-1000 dark:text-dark-neutral-900 grow [&]:text-200 md:[&]:text-[1.125rem] line-clamp-3 "
    >
      test rating item
    </h3>
    test content
  </a>
</div>
`;

exports[`RatingItem tests is a basic snapshot 1`] = `
<div
  class="flex items-center gap-4 relative py-2 lg:p-3 [&:nth-child(-n+9)]:border-b-1 [&:nth-child(-n+9)]:border-neutral-300 dark:[&:nth-child(-n+9)]:border-dark-neutral-350 sm:[&:nth-last-child(2)]:border-none transition-colors duration-150 hover:bg-neutral-100 active:bg-neutral-0 dark:hover:bg-dark-neutral-200 dark:active:bg-dark-neutral-250  md:[counter-increment:list-item_1] before:hidden md:before:inline-block before:content-[counter(list-item,decimal-leading-zero)] before:text-500 lg:before:text-700 before:font-bold before:text-neutral-1000 dark:before:text-dark-neutral-900 before:leading-none before:transition-colors before:duration-150 "
  data-testid="test-rating-item"
>
  <a
    class="grow flex items-center gap-3 before:absolute before:w-full before:h-full before:top-0 before:left-0"
    href="/movies/0"
    title="test rating item"
  >
    <img
      alt="test rating item"
      class="w-[50px] h-[75px] lg:w-[60px] lg:h-[90px] aspect-[2/3]"
      data-nimg="1"
      decoding="async"
      height="90"
      loading="lazy"
      src="http://localhost/assets/images/movie-card-placeholder-l-v.svg"
      style="color: transparent;"
      width="60"
    />
    <h3
      class="leading-[1.25] -tracking-[0.01em] transition-colors duration-150 text-300 font-bold text-neutral-1000 dark:text-dark-neutral-900 grow [&]:text-200 md:[&]:text-[1.125rem] line-clamp-3 "
    >
      test rating item
    </h3>
    test content
  </a>
</div>
`;
